src/Microsoft.ML.FastTree/FastTree.cs

```diff
@@ -1012,7 +1012,7 @@ private protected bool UsingMaxLabel
         }
 
         private FeatureFlockBase CreateOneHotFlock(IChannel ch,
-                List<int> features, int[] binnedValues, int[] lastOn, ValuesList[] instanceList,
+                List<int> features, int[] binnedValues, int[] lastOn, Dictionary<int, ValuesList> instanceList,
```

Comment:
Nice fix! Being one of the primary trainers in ML․NET, I'd recommend testing the speed & memory on a variety of datasets.

Reply:
I see that you're reporting speed/memory on the FastTree ranking `Test` (not `TrainTest`); [Test_Ranking_MSLRWeb10K_RawNumericFeatures_FastTreeRanking](https://github.com/dotnet/machinelearning/blob/161a32942c7f1d2b95b515739bf6d5bba6ee0885/test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs#L103-L111) benchmark. For MAML, `Test` only runs the prediction step. 

Running the training (`TrainTest`) [TrainTest_Ranking_MSLRWeb10K_RawNumericFeatures_FastTreeRanking](https://github.com/dotnet/machinelearning/blob/161a32942c7f1d2b95b515739bf6d5bba6ee0885/test/Microsoft.ML.Benchmarks/Numeric/Ranking.cs#L35-L47) might be more telling.

The MSLR dataset is purely numeric, no text columns in it; so shouldn't be affected by the ngram length change.

## Adding a FastTree text benchmark
You could add FastTree to the text benchmark in ([Benchmarks/Text/MultiClassClassification.cs](https://github.com/dotnet/machinelearning/blob/161a32942c7f1d2b95b515739bf6d5bba6ee0885/test/Microsoft.ML.Benchmarks/Text/MultiClassClassification.cs)). 

Code for adding a FastTree benchmark on WikiDetox would be::
```
        [Benchmark]
        public void CV_Multiclass_WikiDetox_BigramsAndTrichar_OVAFastTree()
        {
            string cmd = @"CV k=5 data=" + _dataPathWiki +
                        " loader=TextLoader{quote=- sparse=- col=Label:R4:0 col=rev_id:TX:1 col=comment:TX:2 col=logged_in:BL:4 col=ns:TX:5 col=sample:TX:6 col=split:TX:7 col=year:R4:3 header=+}" +
                        " xf=Convert{col=logged_in type=R4}" +
                        " xf=CategoricalTransform{col=ns}" +
                        " xf=TextTransform{col=FeaturesText:comment wordExtractor=NGramExtractorTransform{ngram=2}}" +
                        " xf=Concat{col=Features:FeaturesText,logged_in,ns}" +
                        " tr=OVA{p=FastTree}";

            var environment = EnvironmentFactory.CreateClassificationEnvironment<TextLoader, OneHotEncodingTransformer, AveragedPerceptronTrainer, LinearBinaryModelParameters>();
            cmd.ExecuteMamlCommand(environment);
        }
```

### Heuristic
After testing various datasets, you may find the dictionary is beneficial for small (or large) sparse datasets and if we find that, we could use a heuristic of `useDictionary =  sparseness > 0.95 && rowsTimesSlots < 1E6;`

Reply:
I was talking with Harish this morning about the sparseness and choosing between list/dictionary. We are going to discuss it again, but I coudln't find a way to know ahead of time whether the data would be sparse or dense without the user letting us know. 

The test `FastForestBinaryClassificationTestSummary` uses a small dataset. It has about 1300 unique words and total word count is about 9000. The current fast tree trainer multiplies those 2 values together (which is about 11.5 million) and allocates an array of that size right off of the bat which is where the huge memory usage comes from. It ends up using a very small fraction of that amount during training, but I coudln't figure out how to tell how many it was actually using ahead of time.

Reply:
The benchmark I was suggesting to be added, uses WikiDetox which is a 70MB text dataset.

See my recommended code for `CV_Multiclass_WikiDetox_BigramsAndTrichar_OVAFastTree`, above. 

You should be able to just drop that code into ([Benchmarks/Text/MultiClassClassification.cs](https://github.com/dotnet/machinelearning/blob/161a32942c7f1d2b95b515739bf6d5bba6ee0885/test/Microsoft.ML.Benchmarks/Text/MultiClassClassification.cs)). 

> The test FastForestBinaryClassificationTestSummary uses a small dataset. It has about 1300 unique words and total word count is about 9000. 

Tests are on micro-datasets to test if something has changed. The benchmarks are meant to be real-world datasets, like MSLR and WikiDetox.

Would it be possible to measure the sparsity at runtime? Either before you use the array/dictionary on a subsample, or after N% of the data has passed.

Reply:
Ok, after all the investigation, testing, and syncing you and I did, I was only ever able to reproduce this issue when the feature column has been one-hot-encoded. Per our offline discussion, this should never happen (and when it does, usually an error is thrown for trying to make an array that is too large), so the test case appears to be wrong, and my changes are not necessary. Even when the input data was sparse this situation did not repro.

Due to this, I am going to revert my changes to FastTree and just fix the test instead.